File: crates/wallpiperd/src/vk_layer.rs
use crate::config;

pub const VK_CAPTURE_LAYER_NAME: &str = "VK_LAYER_wallpiper_capture";

pub fn write_vk_layer_manifest() {
    let runtime_dir = config::runtime_dir();
    let _ = std::fs::create_dir_all(&runtime_dir);
    let library_path = format!(
        "{}/libVkLayer_wallpiper_capture.so",
        config::install_dir().display()
    );
    let manifest = format!(
        r#"{{
    "file_format_version" : "1.0.0",
    "layer" : {{
        "name": "{VK_CAPTURE_LAYER_NAME}",
        "type": "GLOBAL",
        "library_path": "{library_path}",
        "api_version": "1.1.0",
        "implementation_version": "1",
        "description": "Wallpiper frame capture layer"
    }}
}}
"#
    );
    let path = format!("{runtime_dir}/{VK_CAPTURE_LAYER_NAME}.json");
    if let Err(e) = std::fs::write(&path, manifest) {
        println!("failed to write vk layer manifest at {path}: {e}");
    }
}

File: src/wallpiper-daemon/vk_layer.c
#include "vk_layer.h"

#include <stdio.h>

#include "wallpiper/fsutil.h"
#include "wallpiper/protocol.h"
#include "wallpiper/steam_paths.h"

void wp_write_vk_layer_manifest(void) {
  char temp_dir[512];
  if (!wp_temp_dir(temp_dir, sizeof(temp_dir))) {
    printf("failed to resolve temp dir for vk layer manifest\n");
    return;
  }
  wp_mkdir_p(temp_dir);

  char install_dir[1024];
  if (!wp_install_dir(install_dir, sizeof(install_dir))) {
    printf("failed to resolve install dir for vk layer manifest\n");
    return;
  }

  char path[600];
  int n = snprintf(path, sizeof(path), "%s/%s.json", temp_dir,
                   WP_VK_CAPTURE_LAYER_NAME);
  if (n <= 0 || (size_t)n >= sizeof(path)) {
    printf("vk layer manifest path too long\n");
    return;
  }

  FILE *f = fopen(path, "w");
  if (!f) {
    printf("failed to write vk layer manifest at %s\n", path);
    return;
  }

  fprintf(f,
          "{\n"
          "    \"file_format_version\" : \"1.0.0\",\n"
          "    \"layer\" : {\n"
          "        \"name\": \"%s\",\n"
          "        \"type\": \"GLOBAL\",\n"
          "        \"library_path\": \"%s/libVkLayer_wallpiper_capture.so\",\n"
          "        \"api_version\": \"1.1.0\",\n"
          "        \"implementation_version\": \"1\",\n"
          "        \"description\": \"Wallpiper frame capture layer\"\n"
          "    }\n"
          "}\n",
          WP_VK_CAPTURE_LAYER_NAME, install_dir);
  fclose(f);
}
